### ISSUE TITLE
Train, select, and export the causal transition detector

### ISSUE THREAD

**Issue by jellotheman** [jellotheman/blindsight-prototype#22]
## Parent

- #18 — Stage 3: Causal transition detection

## What to build

Train the two specified causal heads from the cached world states. Calibrate them. Apply the fixed decision policy. Run the three specified evaluations. Select one detector with the fixed rule. Export reproducible ONNX artifacts.

Do not add another unsupervised novelty score. The private v1 evidence already refuted that family for indoor-to-indoor transitions.

## Causal features

Build the 3076-value logistic feature vector at each eligible time step:

- 1024 normalized world-state values;
- 1024 delta values from the prior world state;
- 1024 residual values from the prior exponential moving average;
- cosine to the prior world state;
- cosine to the prior exponential moving average;
- delta norm;
- residual norm.

Use an exponential moving-average factor of 0.1. Update the average one time step at a time. The offline implementation and incremental implementation must return the same values.

## Heads

**Logistic head**

- Standardize the 3076 causal features.
- Fit one linear logit.
- Preserve the fitted mean, scale, weight, and bias.

**Causal GRU head**

- Accept eight ordered world states.
- Map 1024 values to 32 values with a linear layer.
- Apply `tanh`.
- Use a GRU with 32 hidden values.
- Map the final hidden value to one logit.

Both heads return a raw logit. Fit one calibration slope and bias for each head. A calibrated head returns a probability.

## Fixed decision policy

Use these values without post-result tuning:

| Parameter | Value |
| --- | ---: |
| Activate threshold | 0.8 |
| Release threshold | 0.4 |
| Persistence | 2 consecutive high scores |
| Cooldown | 10 seconds |

The policy must update in constant time. It must use only current and past values.

## Evaluation protocol

Keep complete clips and source groups intact. Separate model fitting, calibration, operating-point selection, and threshold-transfer measurement. A group used to fix the operating point cannot provide the final transfer measurement.

Run both window configurations for all three runs:

| Run | Train set | Held-out set | Purpose |
| --- | --- | --- | --- |
| A | Ego4D and HouseTours | Ego4D | Deciding result |
| B | Ego4D and HouseTours | HouseTours | Capture-style check |
| C | HouseTours only | Ego4D | Domain-shift measurement |

Report these values for every run and window configuration:

- recall at one false trigger per ten minutes or fewer;
- median delay for detected proxy transitions;
- recall for each boundary family;
- indoor-to-indoor recall as a separate deciding family;
- average precision for each head;
- held-out group wins for each head;
- false-trigger rate at the fixed threshold;
- calibration quality;
- clip, source-group, positive, ignored, and negative counts.

## Selection and kill rules

Select the GRU only when both conditions are true:

1. Its average precision is at least 0.02 greater than the logistic result.
2. It wins on more than half of the held-out groups.

Select the logistic head in every other case.

Run A decides the stage. The selected detector must reach at least 25 percent recall at one false trigger per ten minutes or fewer. Its median detection delay must be four seconds or less.

If Run A is below the recall floor, stop. Publish the result as a negative result. Do not change a threshold, split, window, label, or selection rule to obtain a pass.

## Artifacts

Export both trained heads with their calibration to ONNX. Create one selected-detector bundle. Record the model version, training manifest, cache revision, hyperparameters, random seeds, selected head, decision policy, and metrics.

The ONNX result must match the training implementation within a declared numerical tolerance. The bundle must not require PyTorch during inference.

## Acceptance criteria

- [ ] Tests prove that offline and incremental feature generation return the same values.
- [ ] Tests prove that feature generation uses no future world state.
- [ ] Training uses complete source groups and the frozen manifest.
- [ ] Calibration data is separate from the final threshold-transfer measurement.
- [ ] Both heads train for both window configurations.
- [ ] Runs A, B, and C use the specified train and held-out corpora.
- [ ] The report contains every required value for each run and configuration.
- [ ] The report names missing or failed traces and explains their treatment.
- [ ] The selection rule is executable and has tests for all outcomes.
- [ ] The decision policy has tests for activation, persistence, release, cooldown, and repeated low scores.
- [ ] Both calibrated heads export to ONNX.
- [ ] ONNX probabilities match the training implementation within tolerance.
- [ ] The selected bundle contains no video or licensed data.
- [ ] A failed kill criterion produces a complete negative report and starts no serving work.
- [ ] The default Stage 0/1 test suite passes without a changed expectation.

## Archived context

- Consult [private issue #24](https://github.com/jellotheman/blind-sight/issues/24) for the cross-building failure, boundary-family results, false-trigger measurement, and cost evidence.
- Consult [private issue #27](https://github.com/jellotheman/blind-sight/issues/27) for the fourteen refuted detector configurations and the causal-alignment finding.
- Consult [private issue #26](https://github.com/jellotheman/blind-sight/issues/26) as unresolved product-risk context. The public specification deliberately accepts room boundaries as proxy labels. Do not silently treat them as product truth.
- Consult private commit [`22a937b`](https://github.com/jellotheman/blind-sight/commit/22a937b395b34ce6f848ead4bedf18f33288f4b7) for the logistic head, GRU head, calibration wrapper, decision policy, grouped training, replay, and artifact shapes. Correct any difference from the public specification.

## Out of scope

- A new unsupervised detector.
- Optical-flow gating.
- A changed operating point.
- Continuous production serving.
- Reference-client behavior.

## Blocked by

- #21 — Cache causal world states on Modal.

**Comment by jellotheman** [comment]
Implemented the training/detector/evaluation slice of this issue in commit 5066850, in three new files: `blindsight/transition/features.py`, `blindsight/transition/detector.py`, `train_transition_detector.py`. Per ADR-0004, this builds and reports **Run A only** (train Ego4D, fix the operating point on a disjoint `heldout` group, measure transfer on a disjoint `test` group) — Runs B and C need HouseTours source video that does not exist.

**Features** (`features.py`): the 3076-value causal vector (1024 normalized + 1024 delta + 1024 residual + 4 scalars, EMA factor 0.1). Two independent implementations: `compute_features_offline` (genuinely vectorized — the EMA recurrence runs through `scipy.signal.lfilter` with an initial condition chosen so `ema[0] == world_state[0]`, not a Python-level replay) and `StreamingFeatureState.step` (true one-step-at-a-time incremental state). Tests assert they agree within float tolerance on random sequences, and that truncating the input array never changes earlier outputs (no future-frame leakage).

**Detector** (`detector.py`):
- `LogisticHead`: mean/scale/weight/bias from `sklearn.LogisticRegression` over standardized features.
- `CausalGRUHead`: linear(1024→32) → tanh → GRU(32) → linear(32→1), history 8, trained by a short Adam/BCE loop with positive-class weighting. Built and tested exactly as specified — but its real parameter count is ~39k, not "about 4000" as the spec's prose estimates; the 1024→32 projection alone is ~32.8k parameters. Flagging this as a likely estimation error in the source doc rather than silently shrinking the architecture.
- `PlattCalibration`: one slope/bias per head, sklearn logistic fit over the raw logit (with a degenerate-single-class fallback).
- `DecisionPolicyState`: activate 0.8 / release 0.4 / persistence 2 / cooldown 10s, O(1) per step. The persistence counter has a dead band between the two thresholds (holds steady rather than resetting) — that's specifically what stops a score dithering near 0.8 from oscillating the count. Tests cover persistence, hysteresis (no oscillation), release-triggered reset, cooldown suppression, and re-arming after cooldown.
- `select_detector_head`: pure function over `(logistic_ap, gru_ap)` per held-out group (grouped by source video) plus overall APs; GRU wins only on ≥0.02 AP margin *and* a strict group-win majority; ties/failures default to logistic. Tested on both branches plus tie and empty-group edge cases.
- ONNX export for both heads with calibration baked into the graph. Had to pin `dynamo=False` (legacy TorchScript exporter) — this torch build's new default dynamo exporter mishandles the dynamic batch axis for these graphs (bakes in batch size 1 despite `dynamic_axes`). Tests confirm ONNX output matches the PyTorch computation within tolerance for both heads.
- `StreamingLogisticDetector`/`StreamingGRUDetector`: the incremental end-to-end path (feature state + head + calibration + policy state) the spec's Artifacts section requires.

**Training script** (`train_transition_detector.py`): takes `--manifest-path`/`--manifest-name`/`--cache-root`/`--window-config` (never hardcoded) plus `--annotations-root`. One gap worth flagging: the frozen manifest records only a boundary *count* per clip, not boundary times, so `label_time_steps` needs the actual `RoomInterval`/`ProxyBoundary` objects from somewhere else — this script re-derives them from the EgoEnv RoomPred CSVs via `corpus.load_egoenv_roompred_directory` + `build_boundary_table` (the same loaders `corpus.py` itself uses), then shifts them into each cache file's local time frame using its `range_start_seconds`. Missing cache files per split are recorded by clip id in the report rather than silently dropped. Reports recall@budget, median delay, per-family recall (indoor-to-indoor called out), AP for both heads, held-out group win counts, achieved false-trigger rate, calibration (Brier score), and clip/positive/ignored/negative counts, plus the kill-criterion check — all computed on `test`.

One shared-contract risk to flag for the encoding side (#21): the clip-id sanitizer in `cache_paths_for_clip`/`sanitize_clip_id` (replace anything outside `[A-Za-z0-9_.-]` with `_`) is my best-effort mirror of what the cache filenames need — I don't own `encode.py` and haven't seen its actual sanitizer, so this needs to be checked for an exact match before a real run.

No real cached world-state files or real EgoEnv annotation CSVs exist in this checkout yet (siblings #20/#21 are still in flight), so everything above is exercised against small synthetic `.npz`/`.json` fixtures and hand-built `RoomInterval`/`ProxyBoundary` rows, including one full CLI smoke test (`test_run_end_to_end_on_synthetic_fixtures`) that runs `train_transition_detector.run()` end to end and checks the report shape and ONNX outputs get written. The real Run A happens once the cache and annotations exist.

`python -m pytest tests/ -q -k transition` and `python -m mypy blindsight/transition/features.py blindsight/transition/detector.py train_transition_detector.py` are both clean (134 passed / 129 deselected; the only other `-k transition`-unrelated failures in the suite are pre-existing `test_transitions.py` failures from the concurrent transition-session-API work, confirmed via `git stash` to be absent on a clean tree and unrelated to this change). Full `pytest tests/` also passes (261 passed, 2 skipped) apart from those same 4 pre-existing failures.

**Comment by jellotheman** [comment]
## Status

Code complete and suite-tested on synthetic fixtures (commit 5066850): ONNX parity with the training implementation, the fixed selection rule, the decision policy, grouped training, and replay-equivalent incremental features are all covered by tests. Never yet run on real cached world states; the world-state cache itself has never been produced by a real encode (noted in the #21 close comment).

Implementation updates since then:

- Measured `wall_seconds` is now recorded in encode run records (49dd108), closing the wall-time gap the #21 close comment called out. GPU-time and retained-byte figures remain estimates; only the live run measures them.
- The spec's code-layout section is reconciled to name both `modal_transition.py` and `modal_transition_encode.py` (49dd108).
- Serving-side hardening landed under #23 (beb151f, ab7e278); it does not affect this ticket's scope.

## Eligibility to close

1. Produce the real cache, cost-guarded, for **both** window configs against the accepted manifest (`ego4d-30gb-v1`). Review the `plan` output's cost/storage estimate before running:
   ```
   modal run modal_transition_encode.py --action plan --manifest-name ego4d-30gb-v1 --window-config short
   modal run modal_transition_encode.py --action run  --manifest-name ego4d-30gb-v1 --window-config short --max-gpu-seconds <budget> --max-retained-bytes <budget>
   ```
   then the same `plan`/`run` pair with `--window-config long`. Record `wall_seconds` and failed items from each run record.
2. Train and evaluate Run A for both configs (per ADR-0004: Ego4D train / heldout / test only; Runs B and C do not apply — HouseTours is dropped):
   ```
   python train_transition_detector.py --manifest-path <frozen manifest JSON> --manifest-name ego4d-30gb-v1 --cache-root <volume mount of world-states/> --annotations-root <EgoEnv RoomPred CSVs> --window-config short --output-dir <dir>
   ```
   and the same command with `--window-config long`. The script re-derives boundary times from the EgoEnv CSVs (the frozen manifest records boundary counts, not times) and names missing cache files in the report.
3. Apply the kill criterion on Run A: **recall >= 0.25 at <= 1 false trigger per 10 minutes, median delay <= 4 s.** A pass closes this issue. A fail publishes the negative result and closes it as stopped-at-kill-criterion, and blocks #23 from starting any serving work.

No numbers above are measured yet: no real run record, evaluation report, or ONNX bundle from a real corpus exists.
